src/storage/core.rs

```diff
@@ -413,54 +416,80 @@ where
     /// Returns entries with matching key
     /// # Errors
     /// Fails after any disk IO errors.
-    pub async fn read_all(&self, key: impl AsRef<K>) -> Result<Vec<Entry>> {
+    pub async fn read_all(&self, key: impl AsRef<K>) -> Result<ReadResult<Vec<Entry>>> {
         let key = key.as_ref();
         let mut all_entries = Vec::new();
         let safe = self.inner.safe.read().await;
+        let blobs = safe.blobs.read().await;
         let active_blob = safe
             .active_blob
             .as_ref()
             .ok_or_else(Error::active_blob_not_set)?;
-        if let Some(entries) = active_blob.read().await.read_all_entries(key).await? {
+        let read = active_blob.read().await.read_all_entries(key).await?;
+        if read.is_found() {
+            let entries = read.unwrap();
             debug!(
                 "storage core read all active blob entries {}",
                 entries.len()
             );
             all_entries.extend(entries);
+        } else if read.is_deleted() {
+            return Ok(read);
         }
-        let blobs = safe.blobs.read().await;
-        let entries_closed_blobs = blobs
-            .iter_possible_childs(key)
+
+        let mut futures = blobs
+            .iter_possible_childs_rev(key)
             .map(|b| b.1.data.read_all_entries(key))
-            .collect::<FuturesUnordered<_>>();
-        entries_closed_blobs
-            .try_filter_map(future::ok)
-            .try_for_each(|v| {
-                debug!("storage core read all closed blob {} entries", v.len());
-                all_entries.extend(v);
-                future::ok(())
-            })
-            .await?;
-        debug!("storage core read all total {} entries", all_entries.len());
-        Ok(all_entries)
+            .collect::<FuturesOrdered<_>>();
+        while let Some(data) = futures.next().await {
+            let data = data?;
+            if data.is_found() {
+                all_entries.extend(data.unwrap());
+            } else if data.is_deleted() {
```

Comment:
In the current implementation, it is impossible to stop on deletion if there is a record in the BLOB after the deletion marker, because `read_all_entries` do not return deletion marker in the collection.

Example:
- blob1: 
  - record_v1
- blob2 (active): 
  - record_v2
  - deletion_marker
  - record_v3

`read_all_entries` on blob2 will return `Found(vec![record_v3])`. 
Then `read_all_entries` on blob1 will return `Found(vec![record_v1])`.
The result will be: `Found(vec![record_v3, record_v1])`.
Deletion is incorrectly skipped, because there is no way to get the deletion marker from blob1 with the current implementation

Reply:
I see, thank you for clarification